pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+		rg.generateResourceClass()
+		rg.generateResourceArgs()
+		rg.generateResourceState()
+	} else {
+		contract.Assert(rg.fun != nil)
+		rg.generateDatasourceFunc()
+		rg.generateDatasourceArgs()
+		rg.generateDatasourceResult()
+	}
+	rg.generateNestedTypes()
+	rg.closeNamespace()
+
+	return file, nil
+}
+
+func (rg *csharpResourceGenerator) openNamespace() {
+	rg.w.Writefmtln("#nullable enable")
+	rg.w.Writefmtln("")
+	rg.w.Writefmtln("using System.Collections.Immutable;")
+	rg.w.Writefmtln("using System.Threading.Tasks;")
+	rg.w.Writefmtln("using Pulumi.Serialization;")
+	rg.w.Writefmtln("")
+	rg.w.Writefmtln("namespace %s", rg.g.moduleNamespace(rg.mod))
+	rg.w.Writefmtln("{")
+}
+
+func (rg *csharpResourceGenerator) closeNamespace() {
+	rg.w.Writefmtln("}")
+}
+
+func (rg *csharpResourceGenerator) generateResourceClass() {
+	// Write the doc comment, if any.
+	if rg.res.doc != "" {
+		emitCSharpDocComment(rg.w, rg.res.doc, rg.res.docURL, "    ")
+	}
+
+	// Open the class.
+	className := rg.name
+	baseType := "Pulumi.CustomResource"
+	if rg.res.IsProvider() {
+		baseType = "Pulumi.ProviderResource"
+	}
+	rg.w.Writefmtln("    public class %s : %s", className, baseType)
+	rg.w.Writefmtln("    {")
+
+	// Emit all output properties.
+	for _, prop := range rg.res.outprops {
+		// Write the property attribute
+		wireName := prop.name
+		propertyName := csharpPropertyName(prop)
+		propertyType := qualifiedCSharpPropertyType(prop, "Outputs", false /*wrapInput*/)
+
+		// Workaround the fact that provider inputs come back as strings.
+		if rg.res.IsProvider() && !isPrimitiveKind(prop.typ.kind) {
+			propertyType = "string"
+			if prop.optional() {
+				propertyType += "?"
+			}
+		}
+
+		emitCSharpPropertyDocComment(rg.w, prop, "        ")
+		rg.w.Writefmtln("        [Output(\"%s\")]", wireName)
+		rg.w.Writefmtln("        public Output<%s> %s { get; private set; } = null!;", propertyType, propertyName)
+		rg.w.Writefmtln("")
+	}
+	rg.w.Writefmtln("")
+
+	// Emit the class constructor.
+	argsType := rg.res.argst.name
+
+	var argsDefault string
+	if len(rg.res.reqprops) == 0 {
+		// If the number of required input properties was zero, we can make the args object optional.
+		argsDefault = " = null"
+		argsType += "?"
+	}
+
+	optionsType := "CustomResourceOptions"
+	if rg.res.IsProvider() {
+		optionsType = "ResourceOptions"
+	}
+
+	// Write a comment prior to the constructor.
+	rg.w.Writefmtln("        /// <summary>")
+	rg.w.Writefmtln("        /// Create a %s resource with the given unique name, arguments, and options.", className)
+	rg.w.Writefmtln("        ///")
+	rg.w.Writefmtln("        /// <param name=\"name\">The unique name of the resource</param>")
+	rg.w.Writefmtln("        /// <param name=\"args\">The arguments used to populate this resource's properties</param>")
+	rg.w.Writefmtln("        /// <param name=\"options\">A bag of options that control this resource's behavior</param>")
+	rg.w.Writefmtln("        /// </summary>")
```

Comment:
`summary` should always close before `param` tags, otherwise, all the parameter comments are in summary:
![image](https://user-images.githubusercontent.com/1454008/67719396-098a0780-f9d2-11e9-9a2d-91ed58c8f209.png)

Reply:
Hah! Good catch.

Reply:
Note: you might have it in other places too. The generated comments look consistently wrong in this respect.